packages/fuels/tests/harness.rs

```diff
@@ -3847,3 +3847,195 @@ async fn test_gas_forwarded_defaults_to_tx_limit() -> Result<(), Error> {
+        "packages/fuels/tests/test_projects/options/out/debug/options-abi.json"
+    );
+
+    let wallet = launch_provider_and_get_wallet().await;
+
+    let contract_id = Contract::deploy(
+        "tests/test_projects/options/out/debug/options.bin",
+        &wallet,
+        TxParameters::default(),
+        StorageConfiguration::default(),
+    )
+    .await?;
+
+    let contract_instance = MyContractBuilder::new(contract_id.to_string(), wallet.clone()).build();
+
+    let s = TestStruct {
+        option: Some(Address::zeroed()),
+    };
+
+    let e = TestEnum::EnumOption(Some(Address::zeroed()));
+
+    let expected_some_address = Some(Address::zeroed());
+    let response = contract_instance.get_some_address().call().await?;
+
+    assert_eq!(response.value, expected_some_address);
+
+    let expected_some_u64 = Some(10);
+    let response = contract_instance.get_some_u64().call().await?;
+
+    assert_eq!(response.value, expected_some_u64);
+
+    let response = contract_instance.get_some_struct().call().await?;
+    assert_eq!(response.value, Some(s.clone()));
+
+    let response = contract_instance.get_some_enum().call().await?;
+    assert_eq!(response.value, Some(e.clone()));
+
+    let response = contract_instance.get_some_tuple().call().await?;
+    assert_eq!(response.value, Some((s.clone(), e.clone())));
+
+    let expected_none = None;
+    let response = contract_instance.get_none().call().await?;
+
+    assert_eq!(response.value, expected_none);
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_rust_option_can_be_encoded() -> Result<(), Error> {
+    abigen!(
+        MyContract,
+        "packages/fuels/tests/test_projects/options/out/debug/options-abi.json"
+    );
+
+    let wallet = launch_provider_and_get_wallet().await;
+
+    let contract_id = Contract::deploy(
+        "tests/test_projects/options/out/debug/options.bin",
+        &wallet,
+        TxParameters::default(),
+        StorageConfiguration::default(),
+    )
+    .await?;
+
+    let s = TestStruct {
+        option: Some(Address::zeroed()),
+    };
+
+    let e = TestEnum::EnumOption(Some(Address::zeroed()));
+
+    let contract_instance = MyContractBuilder::new(contract_id.to_string(), wallet.clone()).build();
+    let expected_u64 = Some(36);
+    let response = contract_instance
+        .input_primitive(expected_u64)
+        .call()
+        .await?;
+
+    assert!(response.value);
+
+    let expected_struct = Some(s);
+    let response = contract_instance
+        .input_struct(expected_struct)
+        .call()
+        .await?;
+
+    assert!(response.value);
+
+    let expected_enum = Some(e);
+    let response = contract_instance.input_enum(expected_enum).call().await?;
+
+    assert!(response.value);
+
+    let expected_none = None;
+    let response = contract_instance.input_none(expected_none).call().await?;
+
+    assert!(response.value);
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_rust_result_can_be_decoded() -> Result<(), Error> {
```

Comment:
I find myself joining decoding and encoding into one test case, seeing as these are e2e and it is less of a problem to do so. 

A sway function that accepts the argument validates its content, and returns the same argument but constructed by sway code so as not to have the bytecode just return your original encoded bytes right back at you.

But that is up to you. With the new macro, these are going to get a lot shorter, then we wouldn't pay the penalty of test code size but only of the time spent executing -- which still hasn't grown unwieldy -- so I'm ok if it remains the way it is.

Reply:
I have checked the generics examples you did. I like them! However, I would keep these separate as the function names directly describe what is done and the code is a bit _clearer_